### ISSUE TITLE
Experiment: behavior-token motion priors for scenario interaction traces

### ISSUE THREAD

**Issue by ll7** [ll7/robot_sf_ll7#4627]
## Context

The arXiv paper `2606.29148` motivates a useful but speculative research direction: pretraining reusable discrete motion/control priors from large motion corpora, then adapting them to downstream tasks instead of learning every behavior from scratch.

For `robot_sf_ll7`, this should **not** become a near-term roadmap item. It is a low-priority experimental note to preserve the idea for later exploration if the existing scenario, pedestrian-interaction, and RL infrastructure becomes mature enough.

## Research question

Can pedestrian/robot interaction traces in `robot_sf_ll7` be represented as short discrete **behavior tokens** or latent skill codes, and can those tokens be used as a diagnostic prior for scenario grouping, failure analysis, or downstream RL initialization?

## Possible experimental slice

Start with an offline, non-invasive prototype:

1. Collect existing simulation traces from representative scenarios.
2. Encode short trajectory windows into compact descriptors, for example:
   - relative robot-pedestrian geometry,
   - velocity and acceleration changes,
   - TTC margin evolution,
   - clearance evolution,
   - path blocking / yielding / passing patterns,
   - recovery smoothness after near-conflict.
3. Cluster or quantize these windows into a small discrete vocabulary.
4. Inspect whether the resulting tokens correspond to interpretable interaction motifs:
   - yielding,
   - assertive passing,
   - oscillatory negotiation,
   - emergency braking,
   - pedestrian displacement,
   - deadlock / freeze,
   - unsafe cut-in.
5. Use the tokens only for diagnostics at first:
   - scenario similarity,
   - collision-cause grouping,
   - regression-test stratification,
   - coverage reporting.

## Explicit non-goals

- Do not train a new controller yet.
- Do not replace the current safety metrics.
- Do not introduce a large transformer dependency.
- Do not make this part of the main evaluation pipeline.
- Do not claim foundation-model behavior unless supported by evidence.

## Suggested implementation boundary

A minimal future prototype could live behind an experimental namespace, for example:

```text
experiments/behavior_tokens/
```

Possible files:

```text
experiments/behavior_tokens/extract_windows.py
experiments/behavior_tokens/quantize_trace_windows.py
experiments/behavior_tokens/inspect_token_motifs.py
experiments/behavior_tokens/README.md
```

The first version should operate only on saved traces and produce static artifacts such as CSV/JSON summaries and simple plots.

## Acceptance criteria for a first exploratory pass

- [ ] A small offline script can load saved interaction traces.
- [ ] Trajectory windows are converted into documented feature vectors.
- [ ] A simple clustering or quantization method assigns each window to a discrete token ID.
- [ ] The resulting token distribution can be summarized per scenario.
- [ ] At least 5–10 example windows per frequent token can be inspected manually.
- [ ] The README states that this is experimental, low priority, and not part of the validated metric stack.

## Priority

Very low. This should only be considered after the current dissertation-critical scenario, metric, prediction, and evaluation issues are stable.

## References

- arXiv:2606.29148 — generative pretrained controllers / reusable motion priors concept.

**Comment by ll7** [comment]
## Implementation plan for #4627

### Goal

Create a very low-priority, offline-only experimental prototype for discrete behavior-token diagnostics from saved scenario interaction traces. The first slice must remain non-invasive: no controller training, no transformer dependency, no metric-stack replacement, and no integration into benchmark/reporting gates.

### Claim boundary

This is exploratory tooling only. It may help inspect scenario similarity, failure motifs, or regression-test stratification, but it must not be used as validated safety evidence or paper/dissertation claim support.

### Proposed namespace

Add an experimental namespace:

```text
experiments/behavior_tokens/
  README.md
  extract_windows.py
  quantize_trace_windows.py
  inspect_token_motifs.py
  schemas.py
```

Keep these scripts decoupled from the normal benchmark pipeline. They should read existing saved traces and write outputs under `output/experiments/behavior_tokens/...` by default.

### Input contract

Use existing episode JSONL rows that contain trace data, especially rows with:

```text
algorithm_metadata.simulation_step_trace.steps
algorithm_metadata.planner_decision_trace.steps   # optional
metrics
outcome
scenario_id
seed
planner_key / scenario_params.algo
```

The scripts should fail closed or skip with explicit reason when a row has no trace. They must not run new campaigns by default.

### Step 1 — trace-window extraction

Implement:

```bash
uv run python experiments/behavior_tokens/extract_windows.py \
  --episode-jsonl 'output/**/episodes.jsonl' \
  --window-steps 10 \
  --stride-steps 5 \
  --output-jsonl output/experiments/behavior_tokens/windows.jsonl \
  --output-csv output/experiments/behavior_tokens/windows.csv
```

Each window should include stable metadata:

- `window_id`;
- `source_episode_path`;
- `episode_id`, if present;
- `scenario_id`;
- `planner_key` / `algo`;
- `seed`;
- `t_start_s`, `t_end_s`;
- `row_status` / outcome flags;
- `feature_schema_version`.

Initial feature vector should be compact and interpretable:

- nearest pedestrian distance / clearance min/mean/slope;
- relative speed to nearest pedestrian;
- TTC proxy min / slope where derivable;
- robot speed mean/min/max;
- robot acceleration or command-change proxy;
- pedestrian speed change near robot;
- route progress delta if present;
- stop/yield proxy: low robot speed while pedestrian is near;
- oscillation proxy: heading or command sign changes;
- near-conflict recovery proxy: clearance increase after minimum clearance.

If a feature is not derivable, record `null` plus a `missing_features` list. Do not silently insert zeros except for genuinely zero-valued measurements.

### Step 2 — token quantization

Implement:

```bash
uv run python experiments/behavior_tokens/quantize_trace_windows.py \
  --windows-jsonl output/experiments/behavior_tokens/windows.jsonl \
  --num-tokens 12 \
  --output-json output/experiments/behavior_tokens/token_assignments.json \
  --output-csv output/experiments/behavior_tokens/token_assignments.csv
```

Implementation guidance:

- Start with deterministic standardization over finite features.
- Prefer a simple algorithm: k-means / mini-batch k-means if already available in dependencies.
- If sklearn is not available, use a deterministic fallback such as quantile binning over 2–3 high-signal axes or a small NumPy-only k-means implementation.
- Record `quantizer_schema_version`, feature names, normalization statistics, token centers, and random seed.
- Never add a large transformer/model dependency for this slice.

### Step 3 — motif inspection

Implement:

```bash
uv run python experiments/behavior_tokens/inspect_token_motifs.py \
  --windows-jsonl output/experiments/behavior_tokens/windows.jsonl \
  --assignments-csv output/experiments/behavior_tokens/token_assignments.csv \
  --output-dir output/experiments/behavior_tokens/inspection
```

Outputs:

- `token_summary.csv` / `token_summary.json` with counts by scenario, planner, outcome, and row status;
- `frequent_tokens.md` with compact motif descriptions;
- `examples/token_<id>_examples.jsonl` containing 5–10 representative windows per frequent token;
- optional simple plots only if existing project dependencies support them.

Motif labels must be manual/diagnostic, e.g. `yielding_candidate`, `oscillatory_negotiation_candidate`, `deadlock_candidate`. Do not treat token ids as ground-truth labels.

### Step 4 — README and usage boundary

`experiments/behavior_tokens/README.md` should state:

- very low priority exploratory idea;
- inspired by reusable motion/control priors but not a foundation-model claim;
- reads saved traces only;
- outputs are diagnostic artifacts under `output/`;
- not part of validated metrics, release gates, benchmark publication, or dissertation claims;
- no new controller training;
- no safety decision may depend on the tokens.

### Tests

Add tests under:

```text
experiments/behavior_tokens/tests/
```

or, if the repo convention prefers central tests:

```text
tests/experiments/test_behavior_tokens.py
```

Required coverage:

- synthetic episode JSONL with `simulation_step_trace` yields expected window count;
- feature extraction handles missing optional fields without fabricating zeros;
- malformed/no-trace rows are skipped with explicit reason;
- quantizer is deterministic for a fixed seed;
- token assignment output has one token id per valid window;
- motif inspection selects bounded representative examples per token;
- README/metadata carries the experimental claim boundary.

### Verification commands

```bash
uv run ruff check experiments/behavior_tokens tests/experiments/test_behavior_tokens.py
uv run pytest tests/experiments/test_behavior_tokens.py -q
uv run python experiments/behavior_tokens/extract_windows.py --help
uv run python experiments/behavior_tokens/quantize_trace_windows.py --help
uv run python experiments/behavior_tokens/inspect_token_motifs.py --help
```

Optional smoke using existing local traces:

```bash
uv run python experiments/behavior_tokens/extract_windows.py \
  --episode-jsonl 'output/**/episodes.jsonl' \
  --max-episodes 20 \
  --output-jsonl output/experiments/behavior_tokens/smoke/windows.jsonl
uv run python experiments/behavior_tokens/quantize_trace_windows.py \
  --windows-jsonl output/experiments/behavior_tokens/smoke/windows.jsonl \
  --num-tokens 8 \
  --output-json output/experiments/behavior_tokens/smoke/token_assignments.json
uv run python experiments/behavior_tokens/inspect_token_motifs.py \
  --windows-jsonl output/experiments/behavior_tokens/smoke/windows.jsonl \
  --assignments-json output/experiments/behavior_tokens/smoke/token_assignments.json \
  --output-dir output/experiments/behavior_tokens/smoke/inspection
```

### Acceptance criteria

- Offline scripts load saved interaction traces and extract documented window features.
- Quantization assigns deterministic discrete token ids.
- Token distributions are summarized by scenario/planner/outcome.
- Representative windows for frequent tokens are inspectable.
- README clearly marks this as low-priority experimental diagnostic tooling.
- No controller training, no transformer dependency, no benchmark pipeline integration, and no safety/claim promotion.
